File: ARRAY AND HASHMAP/Sudoku_Solver.py
class Solution(object):
    def Solve_Sudoku(self, board):
        def is_valid(num, row, col):
            for i in range(9):
                if board[row][i] == num:
                    return False
                if board[i][col] == num:
                    return False
                if board[3 * (row // 3) + i // 3][3 * (col // 3) + i % 3] == num:
                    return False
            return True
        def backtrack():
            for row in range(9):
                for col in range(9):
                    if board[row][col] == '.':
                        for num in map(str, range(1, 10)):
                            if is_valid(num, row, col):
                                board[row][col] = num
                                if backtrack():
                                    return True
                                board[row][col] = '.'
                        return False
            return True
        backtrack()
board = [
    ["5", "3", ".", ".", "7", ".", ".", ".", "."],
    ["6", ".", ".", "1", "9", "5", ".", ".", "."],
    [".", "9", "8", ".", ".", ".", ".", "6", "."],
    ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
    ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
    ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
    [".", "6", ".", ".", ".", ".", "2", "8", "."],
    [".", ".", ".", "4", "1", "9", ".", ".", "5"],
    [".", ".", ".", ".", "8", ".", ".", "7", "9"]
]
object = Solution()
result = object.Solve_Sudoku(board)
for row in result:
    print(row)

                        

File: Blind75/SudokuSolver.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution{
    bool isValid(vector<vector<char>>& board, int row, int col, int num){
        for(int i = 0; i < 9; i++){
            if(board[row][i] == num){
                return false;
            }
        }
        for(int i = 0; i < 9; i++){
            if(board[i][col] == num){
                return false;
            }
        }
        int rowStart = (row/3)*3 , colStart = (col/3) * 3;
        for(int i = rowStart; i < rowStart + 3; i++){
            for(int j =colStart; j < colStart + 3; j++){
                if(board[i][j] == num){
                    return false;
                }
            }
        }
        return true;

    }
    bool solve(vector<vector<char>>& board){
        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] == '.'){
                    for(char num = '1'; num <= '9'; num++){
                        if(isValid(board, i, j, num)){
                            board[i][j] = num;
                            if(solve(board)){
                                return true;
                            }
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    void solveSudoku(vector<vector<char>>& board){
        solve(board);
    }
};
int main(){
    return 0;
}